ts-node-plugin/index.js

```diff
@@ -0,0 +1,38 @@
+const {transform} = require("../dist");
+
+const ModuleKindCommonJS = 1;
+
+/**
+ * ts-node transpiler plugin
+ */
+function create(createOptions) {
+  const {nodeModuleEmitKind} = createOptions;
+  const {module} = createOptions.service.config.options;
+  return {
+    transpile(input, transpileOptions) {
+      const {fileName} = transpileOptions;
+      const transforms = ["typescript"];
+      if (module === ModuleKindCommonJS || nodeModuleEmitKind === "nodecjs") {
+        transforms.push("imports");
+      }
+      if (fileName.endsWith(".tsx") || fileName.endsWith(".jsx")) {
```

Comment:
@cspotcode can you confirm that I'm understanding this right?
* The right way to detect if the JSX transform should be enabled is by looking at the file extension.
* The right way to detect if the file is TypeScript or JavaScript is to look at the file extension. (Unclear if there's much value in disabling the TS transform for JS files, so maybe I'll just leave it enabled for the sake of simplicity, which the swc integration seems to do.)
* ts-node and TS do not allow JSX syntax inside .js files. (I believe for non-TS projects it's fairly common to do that.)

Reply:
JSX detection:
looks like you are correct: https://github.com/TypeStrong/ts-node/blob/97f9afd046b66a0fe05a7d76e7a32f94b872016f/src/transpilers/swc.ts#L65-L68
However, this might be a case where ts-node should take responsibility and pass this as a flag to transpilers.  `transpileOptions.isJsx` or something like that.

TS vs JS detection:
Embarrassingly, the swc integration doesn't attempt this at the moment, it assumes everything is TS.  Again, perhaps this is a case where ts-node should pass as an `isTs` flag to make things clearer.

Re JSX syntax inside JS files: Pretty sure you are correct, though at a higher level, my philosophy is to support whatever the TS language service supports.  So if projects are using TS's editor tooling, and it is happy to do tooltips, tab complete, etc with JSX inside JS files, perhaps using specific tsconfig flags, then I want to support that.